handsontable/src/plugins/copyPaste/copyableRanges.js

```diff
@@ -0,0 +1,194 @@
+import { arrayEach } from '../../helpers/array';
+import { rangeEach } from '../../helpers/number';
+
+/**
+ * The utils class produces the selection ranges in the `{startRow, startCol, endRow, endCol}` format
+ * based on the current table selection. The CopyPaste plugin consumes that ranges to generate
+ * appropriate data ready to copy to the clipboard.
+ *
+ * @private
+ */
+export class CopyableRangesFactory {
+  /**
+   * @type {CellRange}
+   */
+  #selectedRange;
+  /**
+   * @type {function(): number}
+   */
+  #countRows;
+  /**
+   * @type {function(): number}
+   */
+  #countColumns;
+  /**
+   * @type {function(): number}
+   */
+  #rowsLimit;
+  /**
+   * @type {function(): number}
+   */
+  #columnsLimit;
+  /**
+   * @type {function(): number}
+   */
+  #countColumnHeaders;
+
+  /* eslint-disable jsdoc/require-description-complete-sentence */
+  /**
+   * @param {{
+   *   countRows: function(): number,
+   *   countColumns: function(): number,
+   *   rowsLimit: function(): number,
+   *   columnsLimit: function(): number,
+   *   countColumnHeaders: function(): number
+   * }} dependencies The utils class dependencies.
+   */
+  constructor({ countRows, countColumns, rowsLimit, columnsLimit, countColumnHeaders }) {
+    this.#countRows = countRows;
+    this.#countColumns = countColumns;
+    this.#rowsLimit = rowsLimit;
+    this.#columnsLimit = columnsLimit;
+    this.#countColumnHeaders = countColumnHeaders;
+  }
+  /* eslint-enable jsdoc/require-description-complete-sentence */
+
+  /**
+   * Sets the selection range to be processed.
+   *
+   * @param {CellRange} selectedRange The selection range represented by the CellRange class.
+   */
+  setSelectedRange(selectedRange) {
+    this.#selectedRange = selectedRange;
+  }
+
+  /**
+   * Returns a new coords object within the dataset range (cells) with `startRow`, `startCol`, `endRow`
+   * and `endCol` keys.
+   *
+   * @returns {{startRow: number, startCol: number, endRow: number, endCol: number} | null}
+   */
+  getCellsRange() {
+    if (this.#countRows() === 0 || this.#countColumns() === 0) {
+      return null;
+    }
+
+    const {
+      row: startRow,
+      col: startCol,
+    } = this.#selectedRange.getTopStartCorner();
+    const {
+      row: endRow,
+      col: endCol,
+    } = this.#selectedRange.getBottomEndCorner();
+
+    const finalEndRow = Math.min(endRow, Math.max(startRow + this.#rowsLimit() - 1, startRow));
+    const finalEndCol = Math.min(endCol, Math.max(startCol + this.#columnsLimit() - 1, startCol));
+    const isRangeTrimmed = endRow !== finalEndRow || endCol !== finalEndCol;
```

Comment:
This section of the logic seems to be reused (with minor modifications) all across this file. Do you think this could be reorganized so there's less duplicated code?

Reply:
Sounds good. I've moved the logic to separate method (73d536731cef8188062633a1b667171ba8db3a63);